File: moderate/beat_or_bit.rb
def g2d(a)
  a ^= a >> 4
  a ^= a >> 2
  a ^= a >> 1
  a
end

File.open(ARGV[0]).each_line do |line|
  puts line.split('|').map { |x| g2d(x.to_i(2)) }.join(' | ')
end

File: moderate/beat_or_bit.pl
use strict;

sub g2d {
    my $a = $_[0];
    $a ^= $a >> 4;
    $a ^= $a >> 2;
    $a ^= $a >> 1;
    return ($a);
}
open( INFILE, "<", $ARGV[0] )
  or die("Cannot open file $ARGV[0] for reading: $!");
while ( my $line = <INFILE> ) {
    chomp($line);
    my @r = map { g2d( oct( "0b" . $_ ) ) } split( / \| /, $line );
    printf( "%s\n", join( " | ", @r ) );
}
close(INFILE);
